outline.md

```diff
@@ -16,30 +61,127 @@
+
+| type  | availability                                       | components                           | propagates |
 | ---   | ---                                                | ---                                  | ---        |
 | build | build time, not runtime                            | Unknown                              | no         |
 | link  | build time, also implies run and test time for now | Unknown, probably includes a library | yes        |
 | run   | all of build, link run and test                    | Unknown                              | yes        |
 | test  | test build and run time, maybe normal build        | Unknown                              | no         |
 
-### Things not accounted for:
+* Each *build phase* acts upon a [unified environment](#unified-environment), and consists of:
+  1. applying the appropriate [*environment modifications*](#environment-modifications).
+  1. arbitrary work that modifies the *filesystem*.
 
-* PATHs required by a given phase
-* Components used, probably by phase:
-  * commands? might be covered by paths
-  * interface: headers/modules/definitions that are needed at build but not runtime
-  * libraries: possibly also static vs dynamic
-* propagation: in A->B->C does A depend on C in a given phase
-* Does it have ABI impact?
+##### Build Plan
+* A *build plan* is an ordered list of *build phases*.
+
+### Environments
+*From [Environments](https://spack.readthedocs.io/en/latest/environments.html):*
+> An environment is used to group together a set of specs for the purpose of building, rebuilding and deploying in a coherent fashion.
+
+* A *spack environment*
+* What are they?
+  * How/Can these be serialized?
+  * How do these differ from (spack) modules?
+* Shell session?
+* Shared env vars with common parent process?
+
+## Package Developer Concepts
+
+### Package
+As per [Workflows](https://spack.readthedocs.io/en/latest/workflows.html#package-concrete-spec-installed-package), a *package* is a recipe to build one piece of software, containing:
+1. a mapping of version string to *package source*:
+  * **a package source will always resolve to a specific checksum**, but that checksum may change e.g. if it points to a git branch.
+  * a package source can be *unpacked* into a directory to modify the user's filesystem.
+1. a collection of specs representing the package's dependencies.
+  * these describe the dependencies for all versions of the package.
+    * e.g. these may contain *conditional dependencies*, which are [unified](#unified-constraints) during concretization.
+1. implementations for [build phases](#build-phases).
+1. any [environment modifications](#environment-modifications).
+
+These are always specified in a [`package.py`](#package-py) file in some [repo](#package-repositories).
+
+#### Environment Modifications
+* As per [unified environment](#unified-environment), the process environment can be treated as its own global *filesystem*.
```

Comment:
This one I don't understand completely -- we need to work on this.  The filesystem is definitely a place the build can get inputs from, but it's not the only thing in the process environment -- we need to cover env vars, processes, etc. as well, specifically in terms of OS or language runtime constraints on what can be resolved.

For example, in Spack, we kind of abtract where packages live -- each gets its own prefix, and we don't unify them or anything in the build environment.  Within a process, you can load libraries from all the package prefixes, but you can't have symbol conflicts due to the way `ld.so` works.  In `node.js`, that restriction is looser b/c there is a module tree, not a global symbol namespace.

Reply:
I am aware spack *attempts* to abstract where packages live, but regardless the entire global filesystem is currently an input into a spack build phase without sandboxing. A *directory* is itself a *filesystem*, and we may be able to use this in some cases, but for e.g. python 2.6 reaching out manually to `/usr/lib64/libpython2.6.so`, we need to have a non-local representation of that data.

 > we don't unify them or anything in the build environment
They aren't naively unified, specific build outputs specified by a `package.py` are unified with a chroot for each build phase.

Reply:
> regardless the entire global filesystem is currently an input into a spack build phase without sandboxing.

I would say instead that the global filesystem *can be* an input to a spack build phase without sandboxing, and because we don't have sandboxing we don't know which parts.

That said, we try pretty hard not to bring in anything other than system packages like the runtimes under the compilers. So you can reasonably expect that a given random directly outside the Spack `install_tree` that is not `/usr/[lib|bin|include]` is not an input.

But yeah, the point is valid that we currently miss this part of the provenance.  We *are* working towards including it all, though.

Reply:
I have added much more clarification to this part.